File: docx_parser_converter_python/converters/text/table_to_text.py
"""Table to text converter.

Converts Table elements to plain text in various modes (ASCII, tabs, plain).
"""

from dataclasses import dataclass
from typing import Literal

from converters.text.paragraph_to_text import paragraph_to_text
from models.document.paragraph import Paragraph
from models.document.table import Table
from models.document.table_cell import TableCell
from models.document.table_row import TableRow

# =============================================================================
# Type Definitions
# =============================================================================


TableMode = Literal["ascii", "tabs", "plain", "auto"]


@dataclass
class BorderInfo:
    """Information about which table borders are present."""

    top: bool = False
    bottom: bool = False
    left: bool = False
    right: bool = False
    inside_h: bool = False  # Horizontal lines between rows
    inside_v: bool = False  # Vertical lines between columns

    @property
    def has_any(self) -> bool:
        """Check if any border is present."""
        return any([self.top, self.bottom, self.left, self.right, self.inside_h, self.inside_v])


# =============================================================================
# Cell Content Extraction
# =============================================================================


def cell_to_text(cell: TableCell | None) -> str:
    """Extract text content from a table cell.

    Args:
        cell: TableCell element or None

    Returns:
        Text content of the cell
    """
    if cell is None:
        return ""

    parts = []
    for content in cell.content:
        if isinstance(content, Paragraph):
            text = paragraph_to_text(content)
            if text:
                parts.append(text)

    return "\n".join(parts)


def row_to_text(row: TableRow | None, separator: str = "\t") -> str:
    """Extract text content from a table row.

    Args:
        row: TableRow element or None
        separator: Cell separator string

    Returns:
        Text content of the row
    """
    if row is None:
        return ""

    cells = []
    for cell in row.tc:
        text = cell_to_text(cell)
        cells.append(text)

    return separator.join(cells)


# =============================================================================
# Border Detection
# =============================================================================


def _is_border_visible(border) -> bool:
    """Check if a single border is visible."""
    return border is not None and border.val is not None and border.val not in ("none", "nil")


def detect_borders(table: Table) -> BorderInfo:
    """Detect which borders are present on the table.

    Checks both table-level borders and cell-level borders.

    Args:
        table: Table element

    Returns:
        BorderInfo with flags for each border type
    """
    info = BorderInfo()

    # Check table-level borders
    if table.tbl_pr and table.tbl_pr.tbl_borders:
        borders = table.tbl_pr.tbl_borders
        info.top = _is_border_visible(borders.top)
        info.bottom = _is_border_visible(borders.bottom)
        info.left = _is_border_visible(borders.left)
        info.right = _is_border_visible(borders.right)
        info.inside_h = _is_border_visible(borders.inside_h)
        info.inside_v = _is_border_visible(borders.inside_v)

    # Check cell-level borders if table borders not set
    if not info.has_any and table.tr:
        for row_idx, row in enumerate(table.tr):
            for col_idx, cell in enumerate(row.tc):
                if cell.tc_pr and cell.tc_pr.tc_borders:
                    cb = cell.tc_pr.tc_borders
                    # Top row cells with top border -> table has top border
                    if row_idx == 0 and _is_border_visible(cb.top):
                        info.top = True
                    # Bottom row cells with bottom border -> table has bottom border
                    if row_idx == len(table.tr) - 1 and _is_border_visible(cb.bottom):
                        info.bottom = True
                    # Left column cells with left border -> table has left border
                    if col_idx == 0 and _is_border_visible(cb.left):
                        info.left = True
                    # Right column cells with right border -> table has right border
                    if col_idx == len(row.tc) - 1 and _is_border_visible(cb.right):
                        info.right = True
                    # Any cell with bottom border (not last row) -> inside_h
                    if row_idx < len(table.tr) - 1 and _is_border_visible(cb.bottom):
                        info.inside_h = True
                    # Any cell with top border (not first row) -> inside_h
                    if row_idx > 0 and _is_border_visible(cb.top):
                        info.inside_h = True
                    # Any cell with right border (not last col) -> inside_v
                    if col_idx < len(row.tc) - 1 and _is_border_visible(cb.right):
                        info.inside_v = True
                    # Any cell with left border (not first col) -> inside_v
                    if col_idx > 0 and _is_border_visible(cb.left):
                        info.inside_v = True

    return info


def has_visible_borders(table: Table) -> bool:
    """Check if table has visible borders.

    Args:
        table: Table element

    Returns:
        True if table has visible borders
    """
    return detect_borders(table).has_any


# =============================================================================
# ASCII Box Mode
# =============================================================================


def table_to_ascii(table: Table, border_info: BorderInfo | None = None) -> str:
    """Convert table to ASCII box format with partial border support.

    Args:
        table: Table element
        border_info: Optional border info (detected if not provided)

    Returns:
        ASCII box representation matching actual borders
    """
    if not table.tr:
        return ""

    # Detect borders if not provided
    if border_info is None:
        border_info = detect_borders(table)

    # Extract all cell contents and calculate column widths
    rows_data: list[list[str]] = []
    max_cols = 0

    for row in table.tr:
        cells = []
        for cell in row.tc:
            text = cell_to_text(cell)
            # Replace newlines in cell with space for single-line cells
            text = text.replace("\n", " ")
            cells.append(text)
        rows_data.append(cells)
        max_cols = max(max_cols, len(cells))

    if not rows_data:
        return ""

    # Pad rows to have same number of columns
    for row in rows_data:
        while len(row) < max_cols:
            row.append("")

    # Calculate column widths
    col_widths = []
    for col_idx in range(max_cols):
        max_width = 1  # Minimum width
        for row in rows_data:
            if col_idx < len(row):
                max_width = max(max_width, len(row[col_idx]))
        col_widths.append(max_width)

    # Build ASCII table with partial borders
    lines = []

    # Characters for borders
    h_char = "-"  # Horizontal line character
    corner = "+"

    # Helper to build horizontal line
    def build_h_line(left: bool, right: bool, inside_v: bool) -> str:
        if inside_v:
            inner = corner.join(h_char * (w + 2) for w in col_widths)
        else:
            inner = h_char * (sum(col_widths) + 3 * (max_cols - 1) + 2)
        left_char = corner if left else h_char
        right_char = corner if right else h_char
        return left_char + inner + right_char

    # Helper to build data row
    def build_data_row(row: list[str], left: bool, right: bool, inside_v: bool) -> str:
        cells = []
        for col_idx, cell in enumerate(row):
            width = col_widths[col_idx]
            cells.append(f" {cell.ljust(width)} ")

        if inside_v:
            inner = "|".join(cells)
        else:
            inner = " ".join(cells)

        left_char = "|" if left else " "
        right_char = "|" if right else " "
        return left_char + inner + right_char

    # Top border (only if has top border)
    if border_info.top:
        lines.append(build_h_line(border_info.left, border_info.right, border_info.inside_v))

    # Data rows
    for row_idx, row in enumerate(rows_data):
        # Row content
        lines.append(build_data_row(row, border_info.left, border_info.right, border_info.inside_v))

        # Row separator (between rows) - only if has inside_h border
        if row_idx < len(rows_data) - 1 and border_info.inside_h:
            lines.append(build_h_line(border_info.left, border_info.right, border_info.inside_v))

    # Bottom border (only if has bottom border)
    if border_info.bottom:
        lines.append(build_h_line(border_info.left, border_info.right, border_info.inside_v))

    return "\n".join(lines)


# =============================================================================
# Tab-Separated Mode
# =============================================================================


def table_to_tabs(table: Table) -> str:
    """Convert table to tab-separated format.

    Args:
        table: Table element

    Returns:
        Tab-separated representation
    """
    if not table.tr:
        return ""

    lines = []
    for row in table.tr:
        line = row_to_text(row, separator="\t")
        lines.append(line)

    return "\n".join(lines)


# =============================================================================
# Plain Text Mode
# =============================================================================


def table_to_plain(table: Table) -> str:
    """Convert table to plain text format.

    Args:
        table: Table element

    Returns:
        Plain text representation
    """
    if not table.tr:
        return ""

    lines = []
    for row in table.tr:
        line = row_to_text(row, separator="  ")
        lines.append(line)

    return "\n".join(lines)


# =============================================================================
# Main Entry Point
# =============================================================================


def table_to_text(
    table: Table | None,
    mode: TableMode = "auto",
) -> str:
    """Convert a Table to text.

    Args:
        table: Table element or None
        mode: Table rendering mode (ascii, tabs, plain, auto)

    Returns:
        Text representation of the table
    """
    if table is None:
        return ""

    if not table.tr:
        return ""

    # Determine actual mode for auto
    actual_mode = mode
    if mode == "auto":
        actual_mode = "ascii" if has_visible_borders(table) else "tabs"

    # Convert based on mode
    if actual_mode == "ascii":
        if mode == "ascii":
            # Explicit ascii mode: use full borders
            full_borders = BorderInfo(
                top=True, bottom=True, left=True, right=True, inside_h=True, inside_v=True
            )
            return table_to_ascii(table, full_borders)
        else:
            # Auto mode with borders: use detected partial borders
            return table_to_ascii(table)
    elif actual_mode == "tabs":
        return table_to_tabs(table)
    else:  # plain
        return table_to_plain(table)


# =============================================================================
# Table to Text Converter Class
# =============================================================================


class TableToTextConverter:
    """Converter for Table elements to plain text."""

    mode: TableMode

    def __init__(
        self,
        *,
        mode: TableMode = "auto",
    ) -> None:
        """Initialize table converter.

        Args:
            mode: Table rendering mode (ascii, tabs, plain, auto)
        """
        self.mode = mode

    def convert(self, table: Table | None) -> str:
        """Convert a Table to text.

        Args:
            table: Table element or None

        Returns:
            Text representation
        """
        return table_to_text(table, mode=self.mode)

File: docx_parser_converter_ts/src/converters/text/table-to-text.ts
/**
 * Table to text converter.
 *
 * Converts Table elements to plain text in various modes (ASCII, tabs, plain).
 */

import type { Table, TableCell, TableRow } from '../../models/document/table';
import type { Paragraph } from '../../models/document/paragraph';
import type { Border } from '../../models/common/border';
import { paragraphToText } from './paragraph-to-text';

// =============================================================================
// Type Definitions
// =============================================================================

export type TableMode = 'ascii' | 'tabs' | 'plain' | 'auto';

/**
 * Information about which table borders are present.
 */
export interface BorderInfo {
  top: boolean;
  bottom: boolean;
  left: boolean;
  right: boolean;
  insideH: boolean;
  insideV: boolean;
}

/**
 * Create a BorderInfo with default values.
 */
export function createBorderInfo(overrides?: Partial<BorderInfo>): BorderInfo {
  return {
    top: false,
    bottom: false,
    left: false,
    right: false,
    insideH: false,
    insideV: false,
    ...overrides,
  };
}

/**
 * Check if any border is present.
 */
export function hasAnyBorder(info: BorderInfo): boolean {
  return info.top || info.bottom || info.left || info.right || info.insideH || info.insideV;
}

// =============================================================================
// Cell Content Extraction
// =============================================================================

/**
 * Extract text content from a table cell.
 *
 * @param cell - TableCell element or null
 * @returns Text content of the cell
 */
export function cellToText(cell: TableCell | null | undefined): string {
  if (!cell) {
    return '';
  }

  const parts: string[] = [];

  if (cell.content) {
    for (const content of cell.content) {
      if (content && typeof content === 'object' && ('pPr' in content || 'content' in content)) {
        const text = paragraphToText(content as Paragraph);
        if (text) {
          parts.push(text);
        }
      }
    }
  }

  return parts.join('\n');
}

/**
 * Extract text content from a table row.
 *
 * @param row - TableRow element or null
 * @param separator - Cell separator string
 * @returns Text content of the row
 */
export function rowToText(row: TableRow | null | undefined, separator = '\t'): string {
  if (!row) {
    return '';
  }

  const cells: string[] = [];

  if (row.tc) {
    for (const cell of row.tc) {
      const text = cellToText(cell);
      cells.push(text);
    }
  }

  return cells.join(separator);
}

// =============================================================================
// Border Detection
// =============================================================================

/**
 * Check if a single border is visible.
 */
function isBorderVisible(border: Border | null | undefined): boolean {
  return border !== null && border !== undefined &&
         border.val !== null && border.val !== undefined &&
         border.val !== 'none' && border.val !== 'nil';
}

/**
 * Check if a border is explicitly set to none/nil.
 */
function isBorderExplicitlyNone(border: Border | null | undefined): boolean {
  return border !== null && border !== undefined &&
         border.val !== null && border.val !== undefined &&
         (border.val === 'none' || border.val === 'nil');
}

/**
 * Detect which borders are present on the table.
 *
 * Cell-level borders override table-level borders in OOXML.
 * Checks cell borders first, then falls back to table borders.
 *
 * @param table - Table element
 * @returns BorderInfo with flags for each border type
 */
export function detectBorders(table: Table): BorderInfo {
  const info = createBorderInfo();

  if (!table.tr) {
    return info;
  }

  const numRows = table.tr.length;

  // Track whether all cells explicitly set borders to none
  let allCellsHaveNoTop = true;
  let allCellsHaveNoBottom = true;
  let allCellsHaveNoLeft = true;
  let allCellsHaveNoRight = true;
  let allCellsHaveNoInsideH = true;
  let allCellsHaveNoInsideV = true;

  // First pass: check all cell-level borders
  for (let rowIdx = 0; rowIdx < numRows; rowIdx++) {
    const row = table.tr[rowIdx];
    if (!row.tc) continue;

    const numCols = row.tc.length;

    for (let colIdx = 0; colIdx < numCols; colIdx++) {
      const cell = row.tc[colIdx];
      const cb = cell.tcPr?.tcBorders;

      if (cb) {
        // Check if cell has visible borders
        if (rowIdx === 0 && isBorderVisible(cb.top)) {
          info.top = true;
        }
        if (rowIdx === numRows - 1 && isBorderVisible(cb.bottom)) {
          info.bottom = true;
        }
        if (colIdx === 0 && isBorderVisible(cb.left)) {
          info.left = true;
        }
        if (colIdx === numCols - 1 && isBorderVisible(cb.right)) {
          info.right = true;
        }
        if (rowIdx < numRows - 1 && isBorderVisible(cb.bottom)) {
          info.insideH = true;
        }
        if (rowIdx > 0 && isBorderVisible(cb.top)) {
          info.insideH = true;
        }
        if (colIdx < numCols - 1 && isBorderVisible(cb.right)) {
          info.insideV = true;
        }
        if (colIdx > 0 && isBorderVisible(cb.left)) {
          info.insideV = true;
        }

        // Track if NOT explicitly none
        if (rowIdx === 0 && !isBorderExplicitlyNone(cb.top)) {
          allCellsHaveNoTop = false;
        }
        if (rowIdx === numRows - 1 && !isBorderExplicitlyNone(cb.bottom)) {
          allCellsHaveNoBottom = false;
        }
        if (colIdx === 0 && !isBorderExplicitlyNone(cb.left)) {
          allCellsHaveNoLeft = false;
        }
        if (colIdx === numCols - 1 && !isBorderExplicitlyNone(cb.right)) {
          allCellsHaveNoRight = false;
        }
        if (rowIdx < numRows - 1 && !isBorderExplicitlyNone(cb.bottom)) {
          allCellsHaveNoInsideH = false;
        }
        if (rowIdx > 0 && !isBorderExplicitlyNone(cb.top)) {
          allCellsHaveNoInsideH = false;
        }
        if (colIdx < numCols - 1 && !isBorderExplicitlyNone(cb.right)) {
          allCellsHaveNoInsideV = false;
        }
        if (colIdx > 0 && !isBorderExplicitlyNone(cb.left)) {
          allCellsHaveNoInsideV = false;
        }
      } else {
        // Cell doesn't define borders - can't override table borders
        if (rowIdx === 0) allCellsHaveNoTop = false;
        if (rowIdx === numRows - 1) allCellsHaveNoBottom = false;
        if (colIdx === 0) allCellsHaveNoLeft = false;
        if (colIdx === numCols - 1) allCellsHaveNoRight = false;
        allCellsHaveNoInsideH = false;
        allCellsHaveNoInsideV = false;
      }
    }
  }

  // If cells already detected visible borders, return early
  if (hasAnyBorder(info)) {
    return info;
  }

  // Fall back to table borders
  const borders = table.tblPr?.tblBorders;
  if (borders) {
    if (!allCellsHaveNoTop) info.top = isBorderVisible(borders.top);
    if (!allCellsHaveNoBottom) info.bottom = isBorderVisible(borders.bottom);
    if (!allCellsHaveNoLeft) info.left = isBorderVisible(borders.left);
    if (!allCellsHaveNoRight) info.right = isBorderVisible(borders.right);
    if (!allCellsHaveNoInsideH) info.insideH = isBorderVisible(borders.insideH);
    if (!allCellsHaveNoInsideV) info.insideV = isBorderVisible(borders.insideV);
  }

  return info;
}

/**
 * Check if table has visible borders.
 *
 * @param table - Table element
 * @returns True if table has visible borders
 */
export function hasVisibleBorders(table: Table): boolean {
  return hasAnyBorder(detectBorders(table));
}

// =============================================================================
// ASCII Box Mode
// =============================================================================

/**
 * Convert table to ASCII box format with partial border support.
 *
 * @param table - Table element
 * @param borderInfo - Optional border info (detected if not provided)
 * @returns ASCII box representation matching actual borders
 */
export function tableToAscii(table: Table, borderInfo?: BorderInfo): string {
  if (!table.tr || table.tr.length === 0) {
    return '';
  }

  // Detect borders if not provided
  const borders = borderInfo ?? detectBorders(table);

  // Extract all cell contents and calculate column widths
  const rowsData: string[][] = [];
  let maxCols = 0;

  for (const row of table.tr) {
    const cells: string[] = [];
    if (row.tc) {
      for (const cell of row.tc) {
        let text = cellToText(cell);
        // Replace newlines in cell with space for single-line cells
        text = text.replace(/\n/g, ' ');
        cells.push(text);
      }
    }
    rowsData.push(cells);
    maxCols = Math.max(maxCols, cells.length);
  }

  if (rowsData.length === 0) {
    return '';
  }

  // Pad rows to have same number of columns
  for (const row of rowsData) {
    while (row.length < maxCols) {
      row.push('');
    }
  }

  // Calculate column widths
  const colWidths: number[] = [];
  for (let colIdx = 0; colIdx < maxCols; colIdx++) {
    let maxWidth = 1; // Minimum width
    for (const row of rowsData) {
      if (colIdx < row.length) {
        maxWidth = Math.max(maxWidth, row[colIdx].length);
      }
    }
    colWidths.push(maxWidth);
  }

  // Build ASCII table with partial borders
  const lines: string[] = [];

  const hChar = '-';
  const corner = '+';

  // Helper to build horizontal line
  const buildHLine = (left: boolean, right: boolean, insideV: boolean): string => {
    let inner: string;
    if (insideV) {
      inner = colWidths.map(w => hChar.repeat(w + 2)).join(corner);
    } else {
      const totalWidth = colWidths.reduce((sum, w) => sum + w, 0) + 3 * (maxCols - 1) + 2;
      inner = hChar.repeat(totalWidth);
    }
    const leftChar = left ? corner : hChar;
    const rightChar = right ? corner : hChar;
    return leftChar + inner + rightChar;
  };

  // Helper to build data row
  const buildDataRow = (row: string[], left: boolean, right: boolean, insideV: boolean): string => {
    const cells: string[] = [];
    for (let colIdx = 0; colIdx < row.length; colIdx++) {
      const width = colWidths[colIdx];
      cells.push(` ${row[colIdx].padEnd(width)} `);
    }

    let inner: string;
    if (insideV) {
      inner = cells.join('|');
    } else {
      inner = cells.join(' ');
    }

    const leftChar = left ? '|' : ' ';
    const rightChar = right ? '|' : ' ';
    return leftChar + inner + rightChar;
  };

  // Top border (only if has top border)
  if (borders.top) {
    lines.push(buildHLine(borders.left, borders.right, borders.insideV));
  }

  // Data rows
  for (let rowIdx = 0; rowIdx < rowsData.length; rowIdx++) {
    const row = rowsData[rowIdx];

    // Row content
    lines.push(buildDataRow(row, borders.left, borders.right, borders.insideV));

    // Row separator (between rows) - only if has insideH border
    if (rowIdx < rowsData.length - 1 && borders.insideH) {
      lines.push(buildHLine(borders.left, borders.right, borders.insideV));
    }
  }

  // Bottom border (only if has bottom border)
  if (borders.bottom) {
    lines.push(buildHLine(borders.left, borders.right, borders.insideV));
  }

  return lines.join('\n');
}

// =============================================================================
// Tab-Separated Mode
// =============================================================================

/**
 * Convert table to tab-separated format.
 *
 * @param table - Table element
 * @returns Tab-separated representation
 */
export function tableToTabs(table: Table): string {
  if (!table.tr || table.tr.length === 0) {
    return '';
  }

  const lines: string[] = [];
  for (const row of table.tr) {
    const line = rowToText(row, '\t');
    lines.push(line);
  }

  return lines.join('\n');
}

// =============================================================================
// Plain Text Mode
// =============================================================================

/**
 * Convert table to plain text format.
 *
 * @param table - Table element
 * @returns Plain text representation
 */
export function tableToPlain(table: Table): string {
  if (!table.tr || table.tr.length === 0) {
    return '';
  }

  const lines: string[] = [];
  for (const row of table.tr) {
    const line = rowToText(row, '  ');
    lines.push(line);
  }

  return lines.join('\n');
}

// =============================================================================
// Main Entry Point
// =============================================================================

/**
 * Convert a Table to text.
 *
 * @param table - Table element or null
 * @param mode - Table rendering mode (ascii, tabs, plain, auto)
 * @returns Text representation of the table
 */
export function tableToText(table: Table | null | undefined, mode: TableMode = 'auto'): string {
  if (!table || !table.tr || table.tr.length === 0) {
    return '';
  }

  // Determine actual mode for auto
  let actualMode: TableMode = mode;
  if (mode === 'auto') {
    actualMode = hasVisibleBorders(table) ? 'ascii' : 'tabs';
  }

  // Convert based on mode
  if (actualMode === 'ascii') {
    if (mode === 'ascii') {
      // Explicit ascii mode: use full borders
      const fullBorders = createBorderInfo({
        top: true,
        bottom: true,
        left: true,
        right: true,
        insideH: true,
        insideV: true,
      });
      return tableToAscii(table, fullBorders);
    } else {
      // Auto mode with borders: use detected partial borders
      return tableToAscii(table);
    }
  } else if (actualMode === 'tabs') {
    return tableToTabs(table);
  } else {
    return tableToPlain(table);
  }
}

// =============================================================================
// Table to Text Converter Class
// =============================================================================

/**
 * Converter for Table elements to plain text.
 */
export class TableToTextConverter {
  mode: TableMode;

  constructor(options?: { mode?: TableMode }) {
    this.mode = options?.mode ?? 'auto';
  }

  /**
   * Convert a Table to text.
   *
   * @param table - Table element or null
   * @returns Text representation
   */
  convert(table: Table | null | undefined): string {
    return tableToText(table, this.mode);
  }
}
